README.md
# sneakers.js ![Bower Version](https://badge.fury.io/bo/sneakers.svg) [![Powered by You](http://sapegin.github.io/powered-by-you/badge.svg)](http://sapegin.github.io/powered-by-you/)

<!--
[![Build Status](https://secure.travis-ci.org/mar10/sneakers.js.svg?branch=master)](https://travis-ci.org/mar10/sneakers.js) 
-->
> JavaScript plugin that simulates tty typing and decryption effect as seen in the 
1992 movie [Sneakers](https://youtu.be/F5bAa6gFvLs).

Inspired by Brian Barto's [no-more-secrets](https://github.com/bartobri/no-more-secrets) tool.

Copyright &copy; 2016 Martin Wendt.
License: MIT.

Click here for a [demo](https://rawgit.com/mar10/sneakers.js/master/demo/index.html):

[![Demo](demo/console.png)](https://rawgit.com/mar10/sneakers.js/master/demo/index.html)


### Usage

Include <a href="https://github.com/mar10/sneakers.js/blob/master/dist/sneakers.min.js">`sneakers.min.js`</a>
and initialize the plugin 
```js
$(".sneakers").sneakers({
	lockSize: true,  // lock current element size before removing content
	done: function() {
		alert("done");  // optionally handle events
	}
});
```

this will remove the content of the target element, then 'type' it in in an 'encrypted'
version, and then 'decrypt' the text:

```html
<p class="sneakers">
The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
</p>
```

Some CSS rules can help to maintain line breaks and prevent wobbling:

```css
.sneakers {
	/* A monospaced font prevents wobbling */
	font-family: Consolas, "Andale Mono", "Courier New", monospace;
	/* Preserve newlines in plain text */
	white-space: pre;
}
</p>
```


### API

Following all options with their defaults:

```js
$(".sneakers").sneakers({
	typeSpeedMs: 4,          // milliseconds per 'keystroke' (0: no typing simulation)
	jumbleDurationMs: 2000,  // duration of jumble phase [ms] (0: no jumble phase)
	jumbleSpeedMs: 35,       // milliseconds per jumble update
	decodeDurationMs: 5000,  // duration of decoding phase [ms] (0: no decryption simulation)
	decodeSpeedMs: 100,      // milliseconds per update
	stopEps: 0.01,           // decode the rest if less than 1% are encrypted
	lockSize: false,         // lock current element size before removing content
	// Events:
	start: $.noop,           // Current text was replaced by encrypted text 
	type: $.noop,            // 'Encrypted' text has been typed to screen
	done: $.noop             // Decrypted text is completely displayed
});
```


CHANGELOG.md
# 1.0.1-0 / Unreleased
  *

# 1.0.0 / 2016-09-26
  * Initial release


test/spec/sneakers.spec.js
( function( $, QUnit ) {

	"use strict";

	var $testCanvas = $( "#testCanvas" );
	var $fixture = null;

	QUnit.module( "sneakers", {
		beforeEach: function() {

			// fixture is the element where your jQuery plugin will act
			$fixture = $( "<div/>" );

			$testCanvas.append( $fixture );
		},
		afterEach: function() {

			// we remove the element to reset our plugin job :)
			$fixture.remove();
		}
	} );

	QUnit.test( "is inside jQuery library", function( assert ) {

		assert.equal( typeof $.fn.sneakers, "function", "has function inside jquery.fn" );
		assert.equal( typeof $fixture.sneakers, "function", "another way to test it" );
	} );

	QUnit.test( "returns jQuery functions after called (chaining)", function( assert ) {
		assert.equal(
			typeof $fixture.sneakers().on,
			"function",
			"'on' function must exist after plugin call" );
	} );

	QUnit.test( "caches plugin instance", function( assert ) {
		$fixture.sneakers();
		assert.ok(
			$fixture.data( "plugin_sneakers" ),
			"has cached it into a jQuery data"
		);
	} );

	QUnit.test( "decrypt only simulation (no typing)", function( assert ) {
		var done  = assert.async(),
			text = "abc";
		
		assert.expect(9);

		$fixture.text(text);

		assert.equal($fixture.text(), text, "text is initialized");

		$fixture.sneakers( {
			typeSpeedMs: 0,
			jumbleDurationMs: 100,
			decodeDurationMs: 100,
			start: function(){
				assert.ok(true, "start event is triggered");
				assert.equal($fixture.text().length, text.length, "start text exists");
				assert.notEqual($fixture.text(), text, "start text is encrypted");
			},
			type: function(){
				assert.ok(true, "type event is triggered");
				assert.equal($fixture.text().length, text.length, "text is typed");
				assert.notEqual($fixture.text(), text, "typed text is encrypted");
			},
			done: function(){
				assert.ok(true, "done event is triggered");
				assert.equal($fixture.text(), text, "Final result ok");
				done();
			}
		} );
	} );

	QUnit.test( "typing only simulation (no decrypting)", function( assert ) {
		var done  = assert.async(),
			text = "abc";
		
		assert.expect(7);

		$fixture.text(text);

		assert.equal($fixture.text(), text, "text is initialized");

		$fixture.sneakers( {
			typeSpeedMs: 1,
			jumbleDurationMs: 0,
			decodeDurationMs: 0,
			start: function(){
				assert.ok(true, "start event is triggered");
				assert.equal($fixture.text(), "", "text is removed on start");
			},
			type: function(){
				assert.ok(true, "type event is triggered");
				assert.equal($fixture.text(), text, "typed text is NOT encrypted");
			},
			done: function(){
				assert.ok(true, "done event is triggered");
				assert.equal($fixture.text(), text, "Final result ok");
				done();
			}
		} );
	} );

	QUnit.test( "type + decrypt simulation", function( assert ) {
		var done  = assert.async(),
			text = "abc";
		
		assert.expect(8);

		$fixture.text(text);

		assert.equal($fixture.text(), text, "text is initialized");

		$fixture.sneakers( {
			typeSpeedMs: 1,
			jumbleDurationMs: 100,
			decodeDurationMs: 100,
			start: function(){
				assert.ok(true, "start event is triggered");
				assert.equal($fixture.text(), "", "text is removed on start");
			},
			type: function(){
				assert.ok(true, "type event is triggered");
				assert.equal($fixture.text().length, text.length, "text is typed");
				assert.notEqual($fixture.text(), text, "typed text is encrypted");
			},
			done: function(){
				assert.ok(true, "done event is triggered");
				assert.equal($fixture.text(), "abc", "Final result ok");
				done();
			}
		} );
	} );
}( jQuery, QUnit ) );


Gruntfile.js
module.exports = function( grunt ) {

	grunt.initConfig( {

		// Import package manifest
		pkg: grunt.file.readJSON( "package.json" ),

		// Banner definitions
		meta: {
			banner: "/*\n" +
				" *  <%= pkg.title || pkg.name %> - v<%= pkg.version %>\n" +
				" *  <%= pkg.description %>\n" +
				" *  <%= pkg.homepage %>\n" +
				" *\n" +
				" *  Made by <%= pkg.author.name %>\n" +
				" *  Under <%= pkg.license %> License\n" +
				" */\n"
		},

		// Concat definitions
		concat: {
			options: {
				banner: "<%= meta.banner %>"
			},
			dist: {
				src: [ "src/sneakers.js" ],
				dest: "dist/sneakers.js"
			}
		},

		// Lint definitions
		jshint: {
			files: [ "src/sneakers.js", "test/**/*" ],
			options: {
				jshintrc: ".jshintrc"
			}
		},

		jscs: {
			src: "src/**/*.js",
			options: {
				config: ".jscsrc"
			}
		},

		// Minify definitions
		uglify: {
			dist: {
				src: [ "dist/sneakers.js" ],
				dest: "dist/sneakers.min.js"
			},
			options: {
				banner: "<%= meta.banner %>"
			}
		},

		// CoffeeScript compilation
		coffee: {
			compile: {
				files: {
					"dist/sneakers.js": "src/sneakers.coffee"
				}
			}
		},

		// karma test runner
		karma: {
			unit: {
				configFile: "karma.conf.js",
				background: true,
				singleRun: false,
				// browsers: [ "PhantomJS", "Firefox" ]
				// browsers: [ "PhantomJS", "Safari" ]
				browsers: [ "PhantomJS" ]
			},

			//continuous integration mode: run tests once in PhantomJS browser.
			travis: {
				configFile: "karma.conf.js",
				singleRun: true,
				browsers: [ "PhantomJS" ]
			}
		},

		// watch for changes to source
		// Better than calling grunt a million times
		// (call 'grunt watch')
		watch: {
			files: [ "src/*", "test/**/*" ],
			tasks: [ "default" ]
		}

	} );

	grunt.loadNpmTasks( "grunt-contrib-concat" );
	grunt.loadNpmTasks( "grunt-contrib-jshint" );
	grunt.loadNpmTasks( "grunt-jscs" );
	grunt.loadNpmTasks( "grunt-contrib-uglify" );
	grunt.loadNpmTasks( "grunt-contrib-coffee" );
	grunt.loadNpmTasks( "grunt-contrib-watch" );
	grunt.loadNpmTasks( "grunt-karma" );

	grunt.registerTask( "travis", [ "jshint", "karma:travis" ] );
	grunt.registerTask( "lint", [ "jshint", "jscs" ] );
	grunt.registerTask( "build", [ "concat", "uglify" ] );
	grunt.registerTask( "default", [ "jshint", "build", "karma:unit:run" ] );
};


src/sneakers.js
/*******************************************************************************
 * sneakers.js plugin.
 *
 * jQuery plugin that simulates a tty typing and decryption effect as seen in
 * the 1992 movie 'Sneakers'.
 *
 * @see https://github.com/mar10/sneakers.js
 *
 * Copyright (c) 2016, Martin Wendt (http://wwWendt.de). Licensed MIT.
 */
;( function( $, window, document, undefined ) {

	"use strict";

	var validChars = "!\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~",
		// + "☺☻♥♦♣♠•◘○◙♂♀♪♫☼►◄↕‼¶§▬↨↑↓→←∟↔▲▼" +
		// + "│┤╡╢╖╕╣║╗╝╜╛┐└┴┬├─┼╞╟╚╔╩╦╠═╬╧╨╤╥╙╘╒╓╫╪┘┌"
		validCharsLen = validChars.length;

	/* Generate an 'encrypted' replacement for c, unless c is whitespace.*/
	function _randomChar(c) {
		return /\s/.test(c) ? c : validChars[Math.floor(validCharsLen * Math.random())];
	}

	/* 'Encrypt' a string. */
	function _encode(s) {
		var res = "";
		for( var i=0, l=s.length; i<l; i++ ) {
			res += _randomChar(s[i]);
		}
		return res;
	}

	// Create the options defaults
	var pluginName = "sneakers",
		defaults = {
			typeSpeedMs: 4,          // milliseconds per 'keystroke' (0: no typing simulation)
			jumbleDurationMs: 2000,  // duration of jumble phase [ms] (0: no jumble phase)
			jumbleSpeedMs: 35,       // milliseconds per jumble update
			decodeDurationMs: 5000,  // duration of decoding phase [ms] (0: no decryption simulation)
			decodeSpeedMs: 100,      // milliseconds per update
			stopEps: 0.01,           // decode the rest if less than 1% are encrypted
			lockSize: false,         // lock current element size before removing content
			// Events:
			start: $.noop,           // Current text was replaced by encrypted text 
			type: $.noop,            // 'Encrypted' text has been typed to screen
			done: $.noop             // Decrypted text is completely displayed
		};

	// The actual plugin constructor
	function Plugin ( element, options ) {
		this.element = element;
		this.$element = $(element);
		this.settings = $.extend( {}, defaults, options );
		this._defaults = defaults;
		this._name = pluginName;
		this.originalText = null;
		this.currentText = null;
		this.decodeStart = null;
		this.init();
	}

	// Avoid Plugin.prototype conflicts
	$.extend( Plugin.prototype, {
		init: function() {
			var self = this;

			// Prepare encoded text and trigger `start` event
			this.encode();
			this.settings.start.call(this.element);

			// 'Type' simulation and trigger `type` event
			this.type(this.currentText, 0).done(function() {
				self.settings.type.call(self.element);
				console.time("decode");
				// Jumble and decrypt phase
				self.decodeStart = Date.now();
				self.decrypt().done(function() {
					// Trigger `done` event
					console.timeEnd("decode");
					self.settings.done.call(self.element);
				});
			});
		},
		/* Move original text into buffer and set 'decrypted' text instead. */
		encode: function() {
			var $el = this.$element,
				opts = this.settings;

			this.originalText = $el.text();
			// Init with encrypted text unless decryption simulation is off
			if( opts.jumbleDurationMs || opts.decodeDurationMs ) {
				this.currentText = _encode(this.originalText);
			} else {
				this.currentText = this.originalText;
			}
			// Set current dimension as inline style, so the element keeps its
			// size when content is removed for typing simulation
			if( this.settings.lockSize ) {
				$el.css({
					height: $el.css("height"),
					width: $el.css("width")
				});
			}
			// Clear text unless typing simulation is off
			$el.text(opts.typeSpeedMs ? "" : this.currentText);
			// Show element (in case it was hidden before)
			$el.show();
		},
		/* Type text to console, character-by-character. */
		type: function(s, i, _dfd, _prevStamp) {
			var delta, nchar,
				opts = this.settings,
				now = Date.now(),
				dfd = _dfd || new $.Deferred(),
				self = this;

			if( opts.typeSpeedMs && i < s.length ) {
				// Main 'performance' issue seems to be that setTimeout doesn't 
				// allow less than ~10ms intervals.
				// We try to compensate by printing multiple characters at once
				nchar = 1;
				if( _prevStamp ) {
					delta = now - _prevStamp;
					nchar = Math.floor(delta / opts.typeSpeedMs);
				}
				this.$element.append(s.substr(i, nchar));
				// Schedule next keystroke
				setTimeout(function(){
					self.type.call(self, s, i+nchar, dfd, now);
				}, opts.typeSpeedMs);
			} else {
				// All characters typed (or typing simulation is off)
				this.$element.text(s);
				dfd.resolve();				
			}
			return _dfd ? null : dfd.promise();
		},
		/* Change current text until every character is 'decrypted'.
		 * This is done in two phases: 'jumble' and 'decode':
		 */
		decrypt: function(_dfd) {
			var c, i,
				opts = this.settings,
				now = Date.now(),
				elapsed = now - this.decodeStart,
				jumblePhase = elapsed < opts.jumbleDurationMs,
				decriptionRate = opts.decodeDurationMs ? (elapsed - opts.jumbleDurationMs) / opts.decodeDurationMs : 0,
				dfd = _dfd || new $.Deferred(),
				self = this,
				numEncrypted = 0,
				ot = this.originalText,
				ct = this.currentText,
				len = ct.length,
				s = "";

			if ( (jumblePhase && opts.jumbleDurationMs > 0) || (!jumblePhase && opts.decodeDurationMs > 0) ) {
				// Replace each character with a new random value
				for( i=0; i<len; i++) {
					c = ct[i];
					if( jumblePhase ) {
						numEncrypted++;
						c = _randomChar(c);
					} else if( c !== ot[i] ) {
						// decryption phase: keep solved characters unchanged.
						// we also increase the probability of a match, so it
						// becomes more likely to finish in the requested time
						if( Math.random() > decriptionRate ) {
							numEncrypted++;
							c = _randomChar(c);
						} else {
							c = ot[i];
						}
					}
					s += c;
				}
				if ( numEncrypted / len < opts.stopEps) {
					// Resolve all if (nearly) everything is decrypted
					this.currentText = ot;
					numEncrypted = 0;
				} else {
					this.currentText = s;
				}
			} else {
				// Resolve all if loop count exceeded
				this.currentText = ot;
			}
			// Print the current text version
			this.$element.text(this.currentText);

			if( numEncrypted > 0 ) {
				// Schedule next decryption iteration
				setTimeout(function(){
					self.decrypt.call(self, dfd);
				}, jumblePhase ? opts.jumbleSpeedMs : opts.decodeSpeedMs);
			} else {
				dfd.resolve();				
			}
			return _dfd ? null : dfd.promise();
		}
	} );

	// A really lightweight plugin wrapper around the constructor,
	// preventing against multiple instantiations
	$.fn[ pluginName ] = function( options ) {
		return this.each( function() {
			if ( !$.data( this, "plugin_" + pluginName ) ) {
				$.data( this, "plugin_" +
					pluginName, new Plugin( this, options ) );
			}
		} );
	};

} )( jQuery, window, document );
